docs/docs/internals/explicit-nulls.md

```diff
@@ -0,0 +1,182 @@
+
+Type Hierarchy
+--------------
+We change the type hierarchy so that `Null` is only a subtype of `Any` by:
+  - modifying the notion of what is a nullable class (`isNullableClass`) in `SymDenotations`
+    to include _only_ `Null` and `Any`
+  - changing the parent of `Null` in `Definitions` to point to `Any` and not `AnyRef`
+  - changing `isBottomType` and `isBottomClass` in `Definitions`
+
+Java Interop
+------------
+The problem we're trying to solve here is: if we see a Java method `String foo(String)`,
+what should that method look like to Scala?
+  - since we should be able to pass `null` into Java methods, the argument type should be `String|JavaNull`
+  - since Java methods might return `null`, the return type should be `String|JavaNull`
+
+`JavaNull` here is a type alias for `Null` with "magic" properties (see below).
+
+At a high-level:
+  - we track the loading of Java fields and methods as they're loaded by the compiler
+  - we do this in two places: `Namer` (for Java sources) and `ClassFileParser` (for bytecode)
+  - whenever we load a Java member, we "nullify" its argument and return types
+
+The nullification logic lives in `JavaNullInterop.scala`, a new file.
+
+The entry point is the function `def nullifyMember(sym: Symbol, tp: Type)(implicit ctx: Context): Type`
+which, given a symbol and its "regular" type, produces what the type of the symbol should be in the
+explicit nulls world.
+
+In order to nullify a member, we first pass it through a "whitelist" of symbols that need
+special handling (e.g. `constructors`, which never return `null`). If none of the "policies" in the
+whitelist apply, we then process the symbol with a `TypeMap` that implements the following nullification
+function `n`:
+  1. n(T)              = T|JavaNull              if T is a reference type
+  2. n(T)              = T                       if T is a value type
+  3. n(C[T])           = C[T]|JavaNull           if C is Java-defined
+  4. n(C[T])           = C[n(T)]|JavaNull        if C is Scala-defined
+  5. n(A|B)            = n(A)|n(B)|JavaNull
+  6. n(A&B)            = n(A) & n(B)
+  7. n((A1, ..., Am)R) = (n(A1), ..., n(Am))n(R) for a method with arguments (A1, ..., Am) and return type R
+  8. n(T)              = T                       otherwise
+
+JavaNull
+--------
+`JavaNull` is just an alias for `Null`, but with magic power. `JavaNull`'s magic (anti-)power is that
+it's unsound.
+
+```scala
+val s: String|JavaNull = "hello"
+s.length // allowed, but might throw NPE
+```
+
+`JavaNull` is defined as `JavaNullAlias` in `Definitions`.
+The logic to allow member selections is defined in `findMember` in `Types.scala`:
+  - if we're finding a member in a type union
+  - and the union contains `JavaNull` on the r.h.s. after normalization (see below)
+  - then we can continue with `findMember` on the l.h.s of the union (as opposed to failing)
+
+Working with Nullable Unions
+----------------------------
```

Comment:
Can we move these to a Decorator in a separate file instead? `Types.scala` is already too long as it is.

Reply:
Done. See https://github.com/lampepfl/dotty/pull/6344/commits/5a1291339d30f9ce40b34ebb8b049cf4d052738d